---
layout: post
title:  "Medaka v2: progress and potential pitfalls"
date:   2024-10-17
author: Ryan Wick
---

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.13943047.svg)](https://doi.org/10.5281/zenodo.13943047)

A new version of [Medaka](https://github.com/nanoporetech/medaka) was recently released, featuring a model designed specifically for bacterial genomes. See [this video](https://www.youtube.com/watch?v=rRFREd8qxgs) and [this video](https://www.youtube.com/watch?v=IB6DmU40NIU) from ONT for an overview of the challenges posed by modified bases and some information on the new Medaka release.

As someone focused on bacterial genome assembly, this piqued my interest! Over the past few years, I've mostly moved away from using Medaka, as it didn't usually improve assemblies from sup-basecalled reads.[^oldposts] Could this new model change things? In this post, I'll share my initial thoughts on the new Medaka model and highlight a key pitfall in genome polishing.




## New Medaka model

In Medaka v2.0.0, you can specify the `r1041_e82_400bps_bacterial_methylation` model or simply use the `--bacteria` flag. This model is flexible with basecalling, supporting v4.2, v4.3 and v5.0 basecalls at both hac and sup speeds.[^models]

To quickly test it, I used two genomes with higher-than-normal error rates in my ONT-only assembly: a _Campylobacter lari_ genome with 18 errors and an _Enterobacter cloacae_ genome with 11 errors. While these numbers may seem low, most of my ONT-only assemblies in 2024 have fewer than five errors. I suspect methylation motifs are behind these higher error rates.

<table style="width: 70%; margin-left: auto; margin-right: auto;">
  <thead>
    <tr>
      <th>Genome</th>
      <th style="text-align: center;">Pre-Medaka errors</th>
      <th style="text-align: center;">Medaka v1.12.1 errors</th>
      <th style="text-align: center;">Medaka v2.0.0 errors</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="white-space: nowrap;"><em>C. lari</em></td>
      <td style="text-align: center;">18</td>
      <td style="text-align: center;">28</td>
      <td style="text-align: center;">2</td>
    </tr>
    <tr>
      <td style="white-space: nowrap;"><em>E. cloacae</em></td>
      <td style="text-align: center;">11</td>
      <td style="text-align: center;">30</td>
      <td style="text-align: center;">10</td>
    </tr>
  </tbody>
</table>

In both cases, the previous release of Medaka made the accuracy worse and the new release made it better! While this is based on just two genomes, it's enough to reignite my interest in Medaka.



## Missing plasmid pitfall

Encouraged by these results, I ran the new Medaka model on some _Staphylococcus aureus_ assemblies from a current project. But to my surprise, Medaka often made the assemblies worse.

This genome has a 2.9 Mbp chromosome and two small plasmids: 4.4 kbp and 3.1 kbp. As I dug into the strange Medaka results, I found that the chromosome and 4.4 kbp plasmid share ~850 bp of sequence with ~80% identity. This normally wouldn't be a problem, but some assemblies were missing that plasmid.[^missing]

Here's an [IGV](https://igv.org/) screenshot of the problematic region in Medaka's `calls_to_draft.bam` file from one such assembly:
<p align="center"><picture><img src="/assets/images/medaka_igv_screenshot.png" alt="IGV screenshot Medaka polishing" width="100%"></picture></p>

In the screenshot, the reads that align cleanly are from the chromosome, but others (at the bottom) are plasmid reads that erroneously aligned to the chromosome since the plasmid was missing. The plasmid reads outnumbered the correct chromosome reads, leading Medaka to introduce over 100 changes – all of them errors.

Lesson learned: **Make sure your assembly is structurally sound before running Medaka.** Missing plasmids can cause havoc during polishing.[^version] I also noticed that circularisation overlap (duplicated sequence at contig ends) sometimes acquired errors during Medaka polishing.[^canu_trim]

To avoid this problem, I recommend trying [Hybracter](https://github.com/gbouras13/hybracter) from George Bouras. It includes special logic to recover small plasmids which may be missing from the long-read assembly.[^hybracter] Alternatively, you could filter out any reads that don't have a high-identity full-length alignment before running Medaka.




## Final thoughts

While I am impressed by the new version of Medaka, if you have short reads available, you can probably skip it and go straight to polishing with [Polypolish](https://github.com/rrwick/Polypolish) and/or [Pypolca](https://github.com/gbouras13/pypolca). But if you don't have short reads, the new Medaka model is worth a try.

After running Medaka, how can you assess whether the polished assembly is better or worse? There's no perfect solution, but I find the mean length of predicted proteins to be a decent metric.[^proteins] 
It's also good practice to manually inspect Medaka's changes.[^compare_assemblies] Ideally, Medaka should make scattered, minimal changes. If you notice clusters of changes in a single region, that's a red flag – not just for Medaka, but for any polisher.




## Footnotes

[^oldposts]: See [this post](https://rrwick.github.io/2023/05/05/ont-only-accuracy-with-r10.4.1.html) and [this post](https://rrwick.github.io/2023/12/18/ont-only-accuracy-update.html) for examples.

[^models]: Other Medaka models are specific to a basecalling version and speed, e.g. `r1041_e82_400bps_sup_v5.0.0`. Presumably this reflects the data that model was trained on, so the new bacterial Medaka model was probably trained on a mix of different basecalling models.

[^missing]: A missing small plasmid is a common problem with long-read assemblies. Sometimes it's due to library-prep bias (see [this paper](https://doi.org/10.1099/mgen.0.000631)), but it can also be a fault of the assembler.

[^version]: This missing plasmid issue isn't specific to the new Medaka version – earlier versions would have encountered the same problem.

[^canu_trim]: Canu contigs usually have circularisation overlap, but the contig headers specify how much should be trimmed, so I wrote [this script](https://github.com/rrwick/Trycycler/blob/main/scripts/canu_trim.py) to automate the trimming. I suspect Canu assemblies cleaned by this script will perform better in Medaka than unprocessed Canu assemblies.

[^hybracter]: Read more about Hybracter in [this paper](https://doi.org/10.1099/mgen.0.001244) and [this blog post](https://microbiologysociety.org/blog/from-custom-pipeline-to-scalable-software-assembling-complete-bacterial-genomes-from-long-and-hybrid-reads-with-hybracter.html).

[^proteins]: Assembly errors can truncate coding sequences, so assemblies with more errors tend to have shorter predicted proteins. So when comparing alternative assemblies, the one with a larger mean protein length is likely better. This metric is more sensitive to indel errors than substitutions, because indels create frameshifts which often lead to premature stop codons. A helper script to measure this is available [here](https://github.com/rrwick/Perfect-bacterial-genome-tutorial/wiki/Assessing-quality-without-a-reference#assessing-with-prodigal).

[^compare_assemblies]: I wrote [this script](https://github.com/rrwick/Perfect-bacterial-genome-tutorial/wiki/Comparing-assemblies) to display pre- and post-polishing differences in a human-readable format.
